beacon/sync/src/main/java/tech/pegasys/teku/beacon/sync/historical/ReconstructHistoricalStatesService.java

```diff
@@ -13,18 +13,144 @@
+
+  private final CombinedChainDataClient chainDataClient;
+  private final Spec spec;
+  private final Optional<String> genesisStateResource;
+  private final StorageUpdateChannel storageUpdateChannel;
+
+  public ReconstructHistoricalStatesService(
+      final StorageUpdateChannel storageUpdateChannel,
+      final CombinedChainDataClient chainDataClient,
+      final Spec spec,
+      final Optional<String> genesisStateResource) {
+    this.storageUpdateChannel = storageUpdateChannel;
+    this.chainDataClient = chainDataClient;
+    this.spec = spec;
+    this.genesisStateResource = genesisStateResource;
+  }
 
   @Override
   protected SafeFuture<?> doStart() {
-    return SafeFuture.COMPLETE;
+    if (genesisStateResource.isEmpty()) {
+      return SafeFuture.failedFuture(
+          new IllegalStateException("Genesis state resource not provided"));
+    }
+
+    final String resource = genesisStateResource.get();
+    final BeaconState genesisState;
+    try {
+      genesisState = ChainDataLoader.loadState(spec, resource);
+    } catch (IOException e) {
+      LOG.error("Failed to load initial state", e);
+      return SafeFuture.failedFuture(
+          new InvalidConfigurationException(
+              "Failed to load initial state from " + resource + ": " + e.getMessage()));
+    }
+
+    final SafeFuture<Optional<Checkpoint>> future = chainDataClient.getInitialAnchor();
+    future
+        .thenAccept(
+            checkpoint -> {
+              if (checkpoint.isEmpty()) {
+                return; // todo confirm skip
+              }
+
+              final UInt64 anchorSlot = checkpoint.get().getEpochStartSlot(spec);
+              applyBlocks(genesisState, 100, anchorSlot);
+            })
+        .finish(LOG::error); // todo ignores return value
+
+    return SafeFuture.COMPLETE; // todo check
+  }
+
+  public void applyBlocks(
+      final BeaconState genesisState, final int dataStorageFrequency, final UInt64 anchorSlot) {
+    Context context =
+        new Context(genesisState, UInt64.ZERO, UInt64.ONE, dataStorageFrequency, anchorSlot);
+    applyNextBlock(context).finish(LOG::error); // todo ignores return value
+  }
+
+  private SafeFuture<Void> applyNextBlock(Context context) {
+    if (context.slot.isGreaterThanOrEqualTo(context.anchorSlot)) {
+      return SafeFuture.COMPLETE;
+    }
+
+    return chainDataClient
+        .getBlockAtSlotExact(context.slot)
+        .thenAccept(
+            block -> {
+              try {
+                if (block.isEmpty()) {
+                  return;
+                }
+
+                context.currentState =
+                    spec.replayValidatedBlock(context.currentState, block.orElseThrow());
+                if (context.checkStoringHistoricalStates()) {
+                  storageUpdateChannel
+                      .onFinalizedState(context.currentState)
+                      .finish(LOG::error); // todo ignores return value
+                }
```

Comment:
I think we'll need to break this `thenAccept` block up so that the future returned from `storageUpdateChannel.onFinalizedState` winds up being returned in a thenCompose - that way if the state fails to store our regeneration will fail and the error propagates back up instead of being logged and ignored.

So you wind up with something of the form:

```
getBlockAtSlotExact(context.slot)
  .thenComposeChecked(block -> { // thenComposeChecked lets us avoid catching thee StateTransitionException
    if (block.isEmpty()) {
        return SafeFuture.COMPLETE;
    }
    context.currentState = spec.replayValidatedBlock(context.currentState, block.get());
    if (context.checkStoringHistoricalStates()) {
      return storageUpdateChannel.onFinalizedState(context.currentState);
    } else {
      return SafeFuture.COMPLETE;
    }
  })
  .thenRun(() -> context.incrementSlot())
  .thenCompose(__ -> applyNextBlock(context));
```

I will almost certainly have some detail of that wrong but hopefully you get the idea. :)